File: packages/qvac-sdk/examples/registry-query.ts
import {
  qvacModelRegistryList,
  qvacModelRegistrySearch,
  qvacModelRegistryFindByEngine,
  qvacModelRegistryFindByName,
  qvacModelRegistryFindByQuantization,
  qvacModelRegistryFindByModelType,
  qvacModelRegistryGetModel,
  close,
  type QvacModelRegistryEntry,
} from "@qvac/sdk";

try {
  console.log("QVAC Model Registry Query Examples\n");

  // List all available models
  console.log("Listing all models in QVAC model registry...");
  const allModels = await qvacModelRegistryList();
  console.log(`   Found ${allModels.length} models in registry\n`);

  // Show first 5 models as sample
  console.log("   Sample models:");
  allModels.slice(0, 5).forEach((model) => {
    console.log(
      `   - ${model.name} (${model.addon}, ${model.engine}, ${formatSize(model.expectedSize)})`,
    );
  });
  console.log();

  // Search with filter
  console.log('Searching for "whisper" models...');
  const whisperModels = await qvacModelRegistrySearch({ filter: "whisper" });
  console.log(`   Found ${whisperModels.length} whisper-related models\n`);

  // Find by engine
  console.log("Finding models by engine (@qvac/embed-llamacpp)...");
  const embedModels = await qvacModelRegistryFindByEngine(
    "@qvac/embed-llamacpp",
  );
  console.log(`   Found ${embedModels.length} embedding models`);
  embedModels.slice(0, 3).forEach((model) => {
    console.log(`   - ${model.name} (${model.quantization})`);
  });
  console.log();

  // Find by quantization
  console.log("Finding Q4_0 quantized models...");
  const q4Models = await qvacModelRegistryFindByQuantization("q4");
  console.log(`   Found ${q4Models.length} Q4 quantized models`);
  q4Models.slice(0, 3).forEach((model) => {
    console.log(`   - ${model.name}`);
  });
  console.log();

  // Find by name
  console.log('Finding models by name containing "llama"...');
  const llamaModels = await qvacModelRegistryFindByName("llama");
  console.log(`   Found ${llamaModels.length} llama models`);
  llamaModels.slice(0, 3).forEach((model) => {
    console.log(`   - ${model.name} (${model.addon})`);
  });
  console.log();

  // Find by model type
  console.log("Finding transcription models by model type...");
  const transcriptionModels = await qvacModelRegistryFindByModelType("whisper");
  console.log(`   Found ${transcriptionModels.length} transcription models`);
  transcriptionModels.slice(0, 3).forEach((model: QvacModelRegistryEntry) => {
    console.log(`   - ${model.name}`);
  });
  console.log();

  // Find LLM models by model type
  console.log("Finding LLM models by model type...");
  const llmModels = await qvacModelRegistryFindByModelType("llm");
  console.log(`   Found ${llmModels.length} LLM models`);
  llmModels.slice(0, 3).forEach((model: QvacModelRegistryEntry) => {
    console.log(`   - ${model.name} (${model.quantization})`);
  });
  console.log();

  // Get specific model (if we have at least one model)
  const sampleModel = allModels[0];
  if (sampleModel) {
    console.log(
      `Getting specific model: ${sampleModel.registrySource}/${sampleModel.registryPath}`,
    );
    const model = await qvacModelRegistryGetModel(
      sampleModel.registryPath,
      sampleModel.registrySource,
    );
    console.log("   Model details:");
    console.log(`   - Name: ${model.name}`);
    console.log(`   - Addon: ${model.addon}`);
    console.log(`   - Engine: ${model.engine}`);
    console.log(`   - Quantization: ${model.quantization}`);
    console.log(`   - Expected size: ${formatSize(model.expectedSize)}`);
    console.log(`   - Checksum: ${model.sha256Checksum.slice(0, 16)}...`);
    console.log();
  }

  console.log("QVAC model registry query examples completed successfully!");

  close();
  process.exit(0);
} catch (error) {
  console.error("Error:", error);
  process.exit(1);
}

function formatSize(bytes: number): string {
  if (bytes < 1024) return `${bytes}B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)}KB`;
  if (bytes < 1024 * 1024 * 1024)
    return `${(bytes / (1024 * 1024)).toFixed(1)}MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(1)}GB`;
}

File: packages/sdk-python/examples/registry_query.py
"""Python port of packages/sdk/examples/registry-query.ts.

Query the QVAC model registry: list, filter, search by engine / quantization /
addon, and fetch one model's full record. These are request-reply methods on
the flat `tetherto.qvac_sdk` surface; no model needs to be loaded, but a running worker (the
transport) is required.

RUN: python examples/registry_query.py
"""

from __future__ import annotations

import asyncio
import sys

from tetherto.qvac_sdk import (
    Client,
    model_registry_get_model,
    model_registry_list,
    model_registry_search,
)


def val(x):
    """Registry enum fields (addon/engine) are pydantic enums; render the wire
    string, matching the JS SDK's plain-string fields."""
    return x.value if hasattr(x, "value") else x


def format_size(n) -> str:
    if n < 1024:
        return f"{n}B"
    if n < 1024**2:
        return f"{n / 1024:.1f}KB"
    if n < 1024**3:
        return f"{n / 1024**2:.1f}MB"
    return f"{n / 1024**3:.1f}GB"


async def main() -> int:
    async with Client() as client:
        t = client.transport
        try:
            print("▸ QVAC Model Registry Query Examples\n")

            print("▸ Listing all models in QVAC model registry...")
            all_models = await model_registry_list(t)
            print(f"▸ Found {len(all_models)} models in registry\n")

            print("▸ Sample models:")
            for m in all_models[:5]:
                print(
                    f"   - {m.name} ({val(m.addon)}, {val(m.engine)}, {format_size(m.expected_size)})"
                )
            print()

            print('▸ Searching for "whisper" models...')
            whisper = await model_registry_search(t, filter="whisper")
            print(f"▸ Found {len(whisper)} whisper-related models\n")

            print("▸ Searching by engine (llamacpp-embedding)...")
            embed_models = await model_registry_search(t, engine="llamacpp-embedding")
            print(f"▸ Found {len(embed_models)} embedding models")
            for m in embed_models[:3]:
                print(f"   - {m.name} ({m.quantization})")
            print()

            print("▸ Searching for Q4 quantized models...")
            q4 = await model_registry_search(t, quantization="q4")
            print(f"▸ Found {len(q4)} Q4 quantized models")
            for m in q4[:3]:
                print(f"   - {m.name}")
            print()

            if all_models:
                sample = all_models[0]
                print(
                    f"▸ Getting specific model: {sample.registry_source}/{sample.registry_path}"
                )
                model = await model_registry_get_model(
                    t, sample.registry_path, sample.registry_source
                )
                print("   Model details:")
                print(f"   - Name: {model.name}")
                print(f"   - Addon: {val(model.addon)}")
                print(f"   - Engine: {val(model.engine)}")
                print(f"   - Quantization: {model.quantization}")
                print(f"   - Expected size: {format_size(model.expected_size)}")
                print()

            print("▸ QVAC model registry query examples completed successfully!")
        except Exception as error:
            print(f"✖ {error}", file=sys.stderr)
            return 1
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
